source/shared/cpp/ObjectModel/MarkDownBlockParser.cpp

```diff
@@ -0,0 +1,417 @@
+#include <iostream>
+#include "MarkDownBlockParser.h"
+
+using namespace AdaptiveCards;
+
+void MarkDownBlockParser::ParseBlock(std::stringstream &stream) 
+{
+    switch(stream.peek())
+    {
+        case '[':
+        {
+            LinkParser linkParser;
+            linkParser.Match(stream);
+            m_parsedResult.Append(linkParser.GetParsedResult());
+            break;
+        }
+        case ']': case ')':
+        {
+            // add these char as token to code gen list
+            m_parsedResult.AddNewTokenToParsedResult(stream.get());
+            break;
+        }
+        default:
+            EmphasisParser emphasisParser;
+            emphasisParser.Match(stream);
+            m_parsedResult.Append(emphasisParser.GetParsedResult());
+    }
+}
+
+// Emphasis Match's syntax is complete when it's Captured
+// Its syntax is always correct, therefore it attemps to
+// capture until it can't capture no more.
+// it moves two states, emphasis state and text state,
+// at each transition of state, one token is captured
+void EmphasisParser::Match(std::stringstream &stream)
+{ 
+    while (m_current_state != Captured)
+    {
+        m_current_state = m_stateMachine[m_current_state](*this, stream, m_current_token);
+    }
+}
+/// captures text untill it see emphasis character. When it does, switch to Emphasis state
+unsigned int EmphasisParser::MatchText(EmphasisParser &parser, std::stringstream &stream, std::string& token)
+{
+    /// MarkDown keywords
+    if (stream.peek() == '[' || stream.peek() == ']' || stream.peek() == ')' || stream.eof())
+    {
+        parser.Flush(token);
+        return Captured;
+    }
+
+    if (parser.IsMarkDownDelimiter(stream.peek()))
+    {
+        // encounterred first emphasis delimiter
+        parser.CaptureCurrentCollectedStringAsRegularToken();
+        DelimiterType emphasisType = EmphasisParser::GetDelimiterTypeForCharAtCurrentPosition(stream.peek()); 
+        parser.UpdateCurrentEmphasisRunState(emphasisType);
+        token += stream.get();
+        return Emphasis;
+    }
+    else
+    {    
+        parser.UpdateLookBehind(stream.peek());
+        token += stream.get();
+        return Text;
+    }
+}
+
+/// captures text untill it see none-emphasis character. When it does, switch to text state
+unsigned int EmphasisParser::MatchEmphasis(EmphasisParser &parser, std::stringstream &stream, std::string& token)
+{
+    if (stream.peek() == '[' || stream.peek() == ']' || stream.peek() == ')' || stream.eof())
+    {
+        parser.Flush(token);
+        return Captured;
```

Comment:
>Captured [](start = 15, length = 8)

This would be easier to read as EmphasisState::Captured

Reply:
(here and throughout)

---
In reply to: [157579979](https://github.com/Microsoft/AdaptiveCards/pull/1045#discussion_r157579979) [](ancestors = 157579979)

Reply:
This hasn't been changed

---
In reply to: [157580904](https://github.com/Microsoft/AdaptiveCards/pull/1045#discussion_r157580904) [](ancestors = 157580904,157579979)

Reply:
sorry, somehow i managed to miss it; it's fixed now

---
In reply to: [157624500](https://github.com/Microsoft/AdaptiveCards/pull/1045#discussion_r157624500) [](ancestors = 157624500,157580904,157579979)